idd/schema/idd_parser.py

```diff
@@ -55,6 +55,48 @@
 
 import re
 
+
+def _squiggle_context(data, problem_index, num_context_lines_around=5):
+    lines = data.file.splitlines()
+    guilty_line_index = 0
+    count = 0
+    squiggle_pos = None
+    for i, line in enumerate(lines):
+        count += len(line) + 1  # +1 for newline
+        if count > problem_index:
+            guilty_line_index = i
+            squiggle_pos = problem_index - (count - len(line)) + 2
+            break
+
+    start_line = max(guilty_line_index - num_context_lines_around, 0)
+    error_lines = lines[start_line : guilty_line_index + num_context_lines_around]
+    squiggly_line = " " * squiggle_pos + "^" + "~" * (40 - squiggle_pos)
+    error_lines.insert(guilty_line_index - start_line + 1, squiggly_line)
+    return "\n".join(error_lines)
+
+
+class IddParsingError(Exception):
+    def __init__(self, message, data, problem_index=None, num_context_lines_around=5):
+        """Constructor for an Exception with a context.
+
+        @param message (str): the message to be printed
+        @param: data (Data): for retrieving the context
+        @param: problem_index (Optional[int]): if not provided, assumes this is the current data.index
+        @param: num_context_lines_around (int): the number of lines to show before and after
+        """
+        if problem_index is None:
+            problem_index = data.index
+        message += "\nContext:\n" + _squiggle_context(
+            data=data, problem_index=problem_index, num_context_lines_around=num_context_lines_around
+        )
+        # Call the base class constructor with the parameters it needs
+        super().__init__(message)
+
+
+class MissingSemiColonException(IddParsingError):
+    pass
```

Comment:
Some error processing helpers, to report common errors with some helpful context.

Example, I'm adding


```
BrokenObject,
  A1, \field Name
  A2, \field Last Field
      \maximum 10
```

and I get 

```
idd_parser.MissingSemiColonException: In object 'BrokenObject', Missing semi-colon in field A2
Context:
  A1 ; \field Version Identifier
      \default 23.2

BrokenObject,
  A1, \field Name
  A2, \field Last Field
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      \maximum 10

SimulationControl,
      \unique-object
ninja: build stopped: subcommand failed.
```

Reply:
:eyes: this is nice :+1: